File: plugins/healthcare/skills/clinical-trial-protocol/scripts/multiplicity_adjustment.py
#!/usr/bin/env python3
"""
Multiplicity Adjustment Calculator for Clinical Trial Design

Adjusts p-values for multiple comparisons using Bonferroni, Holm (step-down),
and Hochberg (step-up) procedures. Reports adjusted p-values, rejection
decisions, and adjusted significance levels.
"""

import argparse
import json
import sys


def bonferroni(p_values, alpha):
    """Bonferroni correction: multiply each p-value by m."""
    m = len(p_values)
    adjusted = [min(p * m, 1.0) for p in p_values]
    rejected = [p <= alpha for p in adjusted]
    return adjusted, rejected


def holm(p_values, alpha):
    """Holm step-down procedure."""
    m = len(p_values)
    indexed = sorted(enumerate(p_values), key=lambda x: x[1])

    adjusted = [0.0] * m
    rejected = [False] * m

    cummax = 0.0
    for rank, (orig_idx, p) in enumerate(indexed):
        adj_p = min(p * (m - rank), 1.0)
        cummax = max(cummax, adj_p)
        adjusted[orig_idx] = cummax

    for i in range(m):
        rejected[i] = adjusted[i] <= alpha

    return adjusted, rejected


def hochberg(p_values, alpha):
    """Hochberg step-up procedure."""
    m = len(p_values)
    indexed = sorted(enumerate(p_values), key=lambda x: x[1], reverse=True)

    adjusted = [0.0] * m
    rejected = [False] * m

    cummin = 1.0
    for rank_from_end, (orig_idx, p) in enumerate(indexed):
        k = rank_from_end  # 0 = largest p
        adj_p = min(p * (k + 1), 1.0)
        cummin = min(cummin, adj_p)
        adjusted[orig_idx] = cummin

    for i in range(m):
        rejected[i] = adjusted[i] <= alpha

    return adjusted, rejected


METHODS = {
    "bonferroni": bonferroni,
    "holm": holm,
    "hochberg": hochberg,
}


def main():
    parser = argparse.ArgumentParser(
        description="Multiplicity Adjustment Calculator",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python multiplicity_adjustment.py --p-values 0.01,0.04,0.03 --method holm
  python multiplicity_adjustment.py --p-values 0.01,0.04,0.03 --method all
        """,
    )
    parser.add_argument("--p-values", required=True,
                        help="Comma-separated p-values")
    parser.add_argument("--method", required=True,
                        choices=["bonferroni", "holm", "hochberg", "all"],
                        help="Adjustment method (or 'all' for all methods)")
    parser.add_argument("--alpha", type=float, default=0.05,
                        help="Family-wise error rate (default: 0.05)")
    parser.add_argument("--labels", type=str,
                        help="Comma-separated hypothesis labels (optional)")
    parser.add_argument("--output", type=str)
    args = parser.parse_args()

    p_values = [float(x) for x in args.p_values.split(",")]
    m = len(p_values)

    if any(p < 0 or p > 1 for p in p_values):
        print("ERROR: All p-values must be between 0 and 1", file=sys.stderr)
        sys.exit(1)

    if args.labels:
        labels = args.labels.split(",")
        if len(labels) != m:
            print("ERROR: Number of labels must match number of p-values",
                  file=sys.stderr)
            sys.exit(1)
    else:
        labels = [f"H{i+1}" for i in range(m)]

    methods = list(METHODS.keys()) if args.method == "all" else [args.method]

    adjustments = {}
    for method_name in methods:
        adjusted, rejected = METHODS[method_name](p_values, args.alpha)
        adjustments[method_name] = {
            "adjusted_p_values": [round(p, 10) for p in adjusted],
            "rejected": rejected,
            "n_rejected": sum(rejected),
        }

    result = {
        "analysis": "multiplicity_adjustment",
        "n_hypotheses": m,
        "alpha": args.alpha,
        "hypotheses": labels,
        "original_p_values": p_values,
        "adjustments": adjustments,
        "assumptions": [
            "P-values are valid (uniformly distributed under H0)",
            f"Family-wise error rate controlled at {args.alpha}",
            "Bonferroni: valid for any dependency structure",
            "Holm: valid for any dependency structure (uniformly more powerful than Bonferroni)",
            "Hochberg: requires non-negative dependency (e.g., independent or positively correlated tests)",
        ],
        "disclaimers": [
            "Multiplicity adjustment is preliminary",
            "Final analysis plan requires biostatistician review",
            "Consider graphical procedures (Bretz et al.) for complex endpoint hierarchies",
        ],
    }

    output_json = json.dumps(result, indent=2)

    if args.output:
        with open(args.output, 'w') as f:
            f.write(output_json)
        print(f"Results written to {args.output}")
    else:
        print(output_json)


if __name__ == "__main__":
    main()

File: plugins/healthcare/skills/clinical-trial-protocol/scripts/multiplicity_adjustment.R
#!/usr/bin/env Rscript
#
# Multiplicity Adjustment Calculator for Clinical Trial Design
#
# Adjusts p-values for multiple comparisons using Bonferroni, Holm (step-down),
# and Hochberg (step-up) procedures.
#

if (!requireNamespace("jsonlite", quietly = TRUE)) {
  message("ERROR: jsonlite package not installed. Run: install.packages('jsonlite')")
  quit(status = 1)
}

bonferroni_adjust <- function(p_values, alpha) {
  m <- length(p_values)
  adjusted <- pmin(p_values * m, 1.0)
  rejected <- adjusted <= alpha
  list(adjusted = adjusted, rejected = rejected)
}

holm_adjust <- function(p_values, alpha) {
  m <- length(p_values)
  ord <- order(p_values)

  adjusted <- numeric(m)
  cummax_val <- 0.0
  for (i in seq_along(ord)) {
    adj_p <- min(p_values[ord[i]] * (m - i + 1), 1.0)
    cummax_val <- max(cummax_val, adj_p)
    adjusted[ord[i]] <- cummax_val
  }

  rejected <- adjusted <= alpha
  list(adjusted = adjusted, rejected = rejected)
}

hochberg_adjust <- function(p_values, alpha) {
  m <- length(p_values)
  ord <- order(p_values, decreasing = TRUE)

  adjusted <- numeric(m)
  cummin_val <- 1.0
  for (i in seq_along(ord)) {
    adj_p <- min(p_values[ord[i]] * i, 1.0)
    cummin_val <- min(cummin_val, adj_p)
    adjusted[ord[i]] <- cummin_val
  }

  rejected <- adjusted <= alpha
  list(adjusted = adjusted, rejected = rejected)
}

parse_args <- function(args) {
  defaults <- list(
    p_values = NULL, method = NULL, alpha = 0.05,
    labels = NULL, output = NULL
  )

  i <- 1
  while (i <= length(args)) {
    key <- args[i]
    if (i + 1 > length(args)) stop(paste("Missing value for argument:", key))
    val <- args[i + 1]

    switch(key,
      "--p-values" = { defaults$p_values <- as.numeric(strsplit(val, ",")[[1]]) },
      "--method" = { defaults$method <- val },
      "--alpha" = { defaults$alpha <- as.numeric(val) },
      "--labels" = { defaults$labels <- strsplit(val, ",")[[1]] },
      "--output" = { defaults$output <- val },
      stop(paste("Unknown argument:", key))
    )
    i <- i + 2
  }

  if (is.null(defaults$p_values)) stop("--p-values is required")
  if (is.null(defaults$method)) stop("--method is required")
  defaults
}

main <- function() {
  args <- parse_args(commandArgs(trailingOnly = TRUE))

  p_values <- args$p_values
  m <- length(p_values)

  if (any(p_values < 0 | p_values > 1)) {
    message("ERROR: All p-values must be between 0 and 1")
    quit(status = 1)
  }

  if (!is.null(args$labels)) {
    labels <- args$labels
    if (length(labels) != m) {
      message("ERROR: Number of labels must match number of p-values")
      quit(status = 1)
    }
  } else {
    labels <- paste0("H", seq_len(m))
  }

  methods_map <- list(
    bonferroni = bonferroni_adjust,
    holm = holm_adjust,
    hochberg = hochberg_adjust
  )

  if (args$method == "all") {
    method_names <- names(methods_map)
  } else {
    method_names <- args$method
  }

  adjustments <- list()
  for (method_name in method_names) {
    res <- methods_map[[method_name]](p_values, args$alpha)
    adjustments[[method_name]] <- list(
      adjusted_p_values = round(res$adjusted, 10),
      rejected = as.logical(res$rejected),
      n_rejected = as.integer(sum(res$rejected))
    )
  }

  result <- list(
    analysis = "multiplicity_adjustment",
    n_hypotheses = as.integer(m),
    alpha = args$alpha,
    hypotheses = labels,
    original_p_values = p_values,
    adjustments = adjustments,
    assumptions = c(
      "P-values are valid (uniformly distributed under H0)",
      paste0("Family-wise error rate controlled at ", args$alpha),
      "Bonferroni: valid for any dependency structure",
      "Holm: valid for any dependency structure (uniformly more powerful than Bonferroni)",
      "Hochberg: requires non-negative dependency (e.g., independent or positively correlated tests)"
    ),
    disclaimers = c(
      "Multiplicity adjustment is preliminary",
      "Final analysis plan requires biostatistician review",
      "Consider graphical procedures (Bretz et al.) for complex endpoint hierarchies"
    )
  )

  json_str <- jsonlite::toJSON(result, auto_unbox = TRUE, pretty = FALSE, digits = NA)
  json_str <- jsonlite::prettify(json_str, indent = 2)

  if (!is.null(args$output)) {
    writeLines(json_str, args$output)
    cat(paste0("Results written to ", args$output, "\n"))
  } else {
    cat(json_str)
    cat("\n")
  }
}

main()
